### PR TITLE
Datamodule cleanup

### PR DESCRIPTION
**Pull request by calebrob6**
Cleaning up how preprocessing in the datamodules works in preparation for writing a prediction utility. 

Now, each DataModule has a `preprocess(self, sample)` method that performs the image preprocessing and doesn't break if there isn't a "mask" or "label" key in the sample. This is important as `predict.py` needs to be able to apply the same image pre-processing that a trained model checkpoint used while training, however the new data that the model is being run over will definitely not have any "mask" or "label" parts.

I also took this opportunity to clean up a few things throughout the DataModules -- I'll go through and explain these changes inline with comments.

### CODE REVIEW

File: conf/etci2021.yaml [hunk 1]
```diff
@@ -7,7 +7,7 @@ experiment:
     encoder_weights: "imagenet"
     learning_rate: 1e-3
     learning_rate_schedule_patience: 6
-    in_channels: 7
+    in_channels: 6
```
Comment (calebrob6) [inline conf/etci2021.yaml:10]: Previously we had 7 channel inputs because we concatenated a given water mask with the image inputs. Generating predictions with a water mask would be tricky at best / outside of the scope of torchgeo so I removed this.

File: torchgeo/datamodules/chesapeake.py [hunk 2]
```diff
@@ -169,24 +169,34 @@ def preprocess(self, sample: Dict[str, Any]) -> Dict[str, Any]:
         Returns:
             preprocessed sample
         """
-        sample["image"] = sample["image"] / 255.0
```
Comment (calebrob6) [inline torchgeo/datamodules/chesapeake.py:172]: I changed all `sample["image"] = sample["image"] / 255.0` to `sample["image"] /= 255.0` as the inline operation shouldn't make any copies.

File: torchgeo/datamodules/chesapeake.py [hunk 3]
```diff
@@ -169,24 +169,34 @@ def preprocess(self, sample: Dict[str, Any]) -> Dict[str, Any]:
...
-        sample["mask"] = sample["mask"].squeeze()
+        sample["image"] = sample["image"].float()
+        sample["image"] /= 255.0
+
+        if "mask" in sample:
```
Comment (calebrob6) [inline torchgeo/datamodules/chesapeake.py:175]: Added this type of check in order to be able to re-use the `preprocess(...)` logic in cases where "masks"/"labels" aren't available.

File: torchgeo/datamodules/chesapeake.py [hunk 4]
```diff
@@ -169,24 +169,34 @@ def preprocess(self, sample: Dict[str, Any]) -> Dict[str, Any]:
...
-            )
-        else:
-            if self.class_set == 5:
-                sample["mask"][sample["mask"] == 5] = 4
-                sample["mask"][sample["mask"] == 6] = 4
-            sample["mask"] = sample["mask"].long()
+        return sample
 
-        sample["image"] = sample["image"].float()
+    def remove_bbox(self, sample: Dict[str, Any]) -> Dict[str, Any]:
```
Comment (calebrob6) [inline torchgeo/datamodules/chesapeake.py:190]: Removing the "bbox" key is required for using the samples in the lightning trainer, however we need the "bbox" for predictions so I decoupled this logic from preprocess.

File: torchgeo/datamodules/cowc.py [hunk 5]
```diff
@@ -42,7 +42,7 @@ def __init__(
         self.batch_size = batch_size
         self.num_workers = num_workers
 
-    def custom_transform(self, sample: Dict[str, Any]) -> Dict[str, Any]:
```
Comment (calebrob6) [inline torchgeo/datamodules/cowc.py:45]: Renamed all image preprocessing to `preprocess(...)`

File: torchgeo/datamodules/eurosat.py [hunk 6]
```diff
@@ -9,7 +9,7 @@
 import pytorch_lightning as pl
 import torch
 from torch.utils.data import DataLoader
-from torchvision.transforms import Compose, Normalize
```
Comment (calebrob6) [inline torchgeo/datamodules/eurosat.py:12]: Removed Normalize as it is more explicit to subtract band means then divide by band stdevs.

File: torchgeo/datamodules/so2sat.py [hunk 7]
```diff
@@ -25,14 +25,6 @@ class So2SatDataModule(pl.LightningDataModule):
 
     band_means = torch.tensor(
         [
-            -3.591224256609313e-05,
```
Comment (calebrob6) [inline torchgeo/datamodules/so2sat.py:28]: Our previous implementation here didn't really make sense. We were loading all band data, doing channel-wise standardization over all bands, then throwing away the S1 bands so I removed loading of the S1 bands.

File: torchgeo/datamodules/ucmerced.py [hunk 8]
```diff
@@ -25,10 +24,6 @@ class UCMercedDataModule(pl.LightningDataModule):
     Uses random train/val/test splits.
     """
 
-    band_means = torch.tensor([0, 0, 0])
```
Comment (calebrob6) [inline torchgeo/datamodules/ucmerced.py:28]: Why

File: torchgeo/datamodules/usavars.py [hunk 9]
```diff
@@ -24,7 +23,6 @@ def __init__(
         self,
         root_dir: str,
         labels: Sequence[str] = USAVars.ALL_LABELS,
-        transforms: Optional[Callable[[Dict[str, Tensor]], Dict[str, Tensor]]] = None,
```
Comment (calebrob6) [inline torchgeo/datamodules/usavars.py:27]: We don't allow transforms to be passed in to any other DataModules

Comment (adamjstewart) [reply torchgeo/datamodules/usavars.py:27]: This is what we call a slippery slope: https://github.com/microsoft/torchgeo/pull/498#discussion_r843194575

Also see https://github.com/microsoft/torchgeo/discussions/596

File: torchgeo/datamodules/eurosat.py (hunk 6 above)
Comment (adamjstewart) [reply torchgeo/datamodules/eurosat.py:12]: Not sure if I agree with this. I think the `Normalize` transform is more explicit since it tells you why you're doing it, not just what you're doing.

File: torchgeo/datamodules/usavars.py [hunk 10]
```diff
@@ -33,18 +31,28 @@ def __init__(
...
+        """Transform a single sample from the Dataset.
+
+        Args:
+            sample: dictionary containing image
+
+        Returns:
+            preprocessed sample
+        """
+        sample["image"] = sample["image"].float()
+        sample["image"] /= 255.0
```
Comment (adamjstewart) [inline torchgeo/datamodules/usavars.py:53]: It seems like we're somewhat arbitrarily deciding whether to normalize to [0, 1] or [-1, 1]. Not sure if there is any advantage to one or the other.

File: conf/etci2021.yaml (hunk 1 above)
Comment (adamjstewart) [reply conf/etci2021.yaml:10]: Wouldn't predictions be out_channels, not in_channels?

File: torchgeo/datamodules/chesapeake.py (hunk 4 above)
Comment (adamjstewart) [reply torchgeo/datamodules/chesapeake.py:190]: Do you remember why we had to remove bbox again? I feel like there was an option to keep it if we did something else differently...

File: conf/etci2021.yaml (hunk 1 above)
Comment (calebrob6) [reply conf/etci2021.yaml:10]: Yes, but the number of in_channels changed from 7 to 6 because we were previously using a pre-computed water mask as a channel.

File: torchgeo/datamodules/chesapeake.py (hunk 4 above)
Comment (calebrob6) [reply torchgeo/datamodules/chesapeake.py:190]: I'm suspicious that it has something to do with collation, but the trainers/test definitely break if you don't remove it.

File: torchgeo/datamodules/eurosat.py (hunk 6 above)
Comment (calebrob6) [reply torchgeo/datamodules/eurosat.py:12]: True, I was thinking of it as, "there are many ways to 'normalize' something where `Normalize` happens to apply channel wise standardization". I don't have a strong opinion here.

File: torchgeo/datamodules/usavars.py (hunk 10 above)
Comment (calebrob6) [reply torchgeo/datamodules/usavars.py:53]: Yep!

File: torchgeo/datamodules/usavars.py (hunk 9 above)
Comment (calebrob6) [reply torchgeo/datamodules/usavars.py:27]: Haha yeah I know; I can't make up my mind here. If we have this `Optional[Callable]` as an argument, then we can't (easily) initialize a class instance from a configuration file. This will make general-purpose scripts like `predict.py` more difficult to write. On the other hand, our datamodules aren't very useful for experimentation as-is, because good train-time augmentations are a very important.

File: torchgeo/datamodules/usavars.py (hunk 10 above)
Comment (calebrob6) [reply torchgeo/datamodules/usavars.py:53]: I wouldn't mind arbitrarily being consistent and do [0,1] normalization for everything.

File: torchgeo/datamodules/eurosat.py (hunk 6 above)
Comment (adamjstewart) [reply torchgeo/datamodules/eurosat.py:12]: I don't have a strong preference here either. I tend to use builtin library functions where possible since I trust them more than I trust my own coding, but this one is obviously very simple. I think the biggest advantage of using `Normalize` is that it does the `.reshape(-1, 1, 1)` here for you, so that's one less thing to worry about. It also uses `.view(-1, 1, 1)` which I think is more efficient. It also accepts a list, so you don't need to convert it to a tensor yourself. Personally I'm gonna vote for using `Normalize` to keep the code as simple as possible.

File: torchgeo/datamodules/usavars.py (hunk 9 above)
Comment (adamjstewart) [reply torchgeo/datamodules/usavars.py:27]: Can't we have a set of default transforms associated with each datamodule? Then users can override them in Python if they need to via this optional callable (experimentation) but we'll still have config files without customization that default to the optimal known transforms (reproduction).

File: torchgeo/datamodules/usavars.py (hunk 10 above)
Comment (adamjstewart) [reply torchgeo/datamodules/usavars.py:53]: I'm not sure if this needs to be solved in this PR, just something to think about. We should probably try both and see if it makes any difference. It may be different on a case-by-case basis.

File: torchgeo/datamodules/usavars.py (hunk 9 above)
Comment (calebrob6) [reply torchgeo/datamodules/usavars.py:27]: Yeah I think that's the best of both worlds. I don't think an "optimal" known transform exists across these datasets -- but a sane default implemented consistently would be a good step. Further, as long as these are optional train transforms, instantiating the DataModule without them shouldn't have any effect on using the DataModule for pre-processing purposes.

**Comment by calebrob6** [comment]
Okay, so I'm going to switch to using Normalize consistently then leave further changes for another PR

**Review by adamjstewart — DISMISSED** [review]
Technically, a lot of these changes (transforms -> preprocess) are API changes that should be documented with `.. versionchanged:: 0.3` but I doubt too many people are calling these directly or overriding them so I won't ask for documentation of these changes.

Inria adds an `augmentations` parameter. Do we want to remove this now for consistency or keep it and think about how to handle this later? Inria was added in this release so its API hasn't been public so we can remove it without documenting the change if we do it now.

**Comment by calebrob6** [comment]
> Inria adds an augmentations parameter. Do we want to remove this now for consistency or keep it and think about how to handle this later? Inria was added in this release so its API hasn't been public so we can remove it without documenting the change if we do it now.

I think remove it

**Review by adamjstewart — APPROVED** [review]
